src/transport.rs

```diff
@@ -0,0 +1,748 @@
+        (self.0.valid & j::JackBBTFrameOffset) != 0
+    }
+
+    /*
+    /// Query to see if the Timecode data is valid.
+    pub fn valid_timecode(&self) -> bool {
+        (self.0.valid & j::JackPositionTimecode) != 0
+    }
+
+    /// Query to see if the Audio/Video ratio is valid.
+    pub fn valid_avr(&self) -> bool {
+        (self.0.valid & j::JackAudioVideoRatio) != 0
+    }
+
+    /// Query to see if the Video frame offset is valid.
+    pub fn valid_video_frame_offset(&self) -> bool {
+        (self.0.valid & j::JackVideoFrameOffset) != 0
+    }
+    */
+
+    /// Get the frame number on the transport timeline.
+    ///
+    /// # Remarks
+    /// * This is not the same as what jack_frame_time returns.
+    pub fn frame(&self) -> Frames {
+        self.0.frame
+    }
+
+    /// Set the frame number on the transport timeline.
+    pub fn set_frame(&mut self, frame: Frames) {
+        self.0.frame = frame;
+    }
+
+    /// Get the current frame rate, in frames per second.
+    ///
+    /// # Remarks
+    /// * This is only set by the server so it will be `None` if this struct hasn't come from the
+    /// sever.
+    pub fn frame_rate(&self) -> Option<Frames> {
+        if self.0.frame_rate > 0 {
+            Some(self.0.frame_rate)
+        } else {
+            None
+        }
+    }
+
+    /// Get a microsecond timestamp.
+    ///
+    /// # Remarks
+    /// * This is only set by the server so it will be `None` if this struct hasn't come from the
+    /// sever.
+    /// * Guaranteed to be monotonic, but not neccessarily linear.
+    /// * The absolute value is implementation-dependent (i.e. it could be wall-clock, time since
+    /// jack started, uptime, etc).
+    pub fn usecs(&self) -> Option<Time> {
+        //NOTE could it actually be 0 and come from the server?
+        if self.0.usecs > 0 {
+            Some(self.0.usecs)
+        } else {
+            None
+        }
+    }
+
+    /// Get the BarBeatsTick data if it is valid.
+    pub fn bbt(&self) -> Option<TransportBBT> {
+        if self.valid_bbt() && self.0.bar > 0 && self.0.beat > 0 && self.0.tick >= 0 {
+            Some(TransportBBT {
+                bar: self.0.bar as _,
+                beat: self.0.beat as _,
+                tick: self.0.tick as _,
+                sig_num: self.0.beats_per_bar,
+                sig_denom: self.0.beat_type,
+                ticks_per_beat: self.0.ticks_per_beat,
+                bpm: self.0.beats_per_minute,
+                bar_start_tick: self.0.bar_start_tick,
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Set the BarBeatsTick data in this position.
+    ///
+    /// # Arguments
+    /// * `bbt` - The data to set in the position. `None` will invalidate the BarBeatsTick data.
+    ///
+    /// # Remarks
+    /// * On failure this will leave the pre-existing data intact.
+    pub fn set_bbt(&mut self, bbt: Option<TransportBBT>) -> std::result::Result<(), TransportBBT> {
```

Comment:
@wmedrano yeah, what do you think of `Error::InvalidTransportBBT(&str reason, TransportBBT)`

Reply:
@wmedrano one issue with adding another error to the enum is that existing match statements have to be updated if they don't have a catch all case. would it be worth adding (non_exhaustive)[https://doc.rust-lang.org/reference/attributes/type_system.html] and doing that, or maybe it would be nicer to just use an error type specific for this and not a catch all error enum for all of jack?

Reply:
Error type specific to this seems fine.